service/controller/resource/monitoring/remotewriteingress/v1/resource.go

```diff
@@ -0,0 +1,146 @@
+		},
+		GetDesiredObject: func(ctx context.Context, v interface{}) (metav1.Object, error) {
+			return toIngress(v, config)
+		},
+		HasChangedFunc: hasChanged,
+	}
+	r, err := generic.New(c)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	return r, nil
+}
+
+func getObjectMeta(v interface{}, config Config) (metav1.ObjectMeta, error) {
+	cluster, err := key.ToCluster(v)
+	if err != nil {
+		return metav1.ObjectMeta{}, microerror.Mask(err)
+	}
+
+	return metav1.ObjectMeta{
+		Name:      fmt.Sprintf("prometheus-%s-remote-write", key.ClusterID(cluster)),
+		Namespace: key.Namespace(cluster),
+		Labels:    key.PrometheusLabels(cluster),
+		Annotations: map[string]string{
+			"nginx.ingress.kubernetes.io/auth-type":   "basic",
+			"nginx.ingress.kubernetes.io/auth-secret": key.RemoteWriteAgentSecretName(),
+			"nginx.ingress.kubernetes.io/auth-realm":  "Authentication Required",
+		},
+	}, nil
+}
+
+func toIngress(v interface{}, config Config) (metav1.Object, error) {
+	if v == nil {
+		return nil, nil
+	}
+
+	objectMeta, err := getObjectMeta(v, config)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	cluster, err := key.ToCluster(v)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	// Note we only configure a path that will cause a location/HTTP path to be
+	// added to the proxy for the base domain here.
+	//
+	// The common server configuration like TLS is configured in the `Ingress`
+	// installed by PMO chart itself. We want to avoid duplicating the TLS
+	// configuration here as only one certificate can be used for a given
+	// domain, and if multiple `Ingress` resources specify that configuration
+	// the Ingress Controller just picks a random one (first one it finds). And
+	// if it picks up a copied certificate, and that copy happens to become out
+	// of date at some point, it would break HTTPS access to that domain.
+	//
+	// So we want TLS configuration to be controlled only by the `Ingress`
+	// resource that also defines the source of the certificates (i.e. the
+	// Let's Encrypt annotation or the static source for the installation)
+	// so we know as soon as it's updated IC will be using it.
+	pathType := networkingv1.PathTypeImplementationSpecific
+	ingressClassName := key.IngressClassName()
+	ingress := &networkingv1.Ingress{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: networkingv1.SchemeGroupVersion.Version,
+			Kind:       "Ingress",
+		},
+		ObjectMeta: objectMeta,
+		Spec: networkingv1.IngressSpec{
+			IngressClassName: &ingressClassName,
+			Rules: []networkingv1.IngressRule{
+				{
+					Host: config.BaseDomain,
+					IngressRuleValue: networkingv1.IngressRuleValue{
+						HTTP: &networkingv1.HTTPIngressRuleValue{
+							Paths: []networkingv1.HTTPIngressPath{
+								{
+									Path: fmt.Sprintf("/%s/api/v1/write", key.ClusterID(cluster)),
+									Backend: networkingv1.IngressBackend{
+										Service: &networkingv1.IngressServiceBackend{
+											Name: "prometheus-operated",
```

Comment:
Can you put the name of service as constant.

Reply:
Sure